### PR TITLE
Add Fishr: Invariant Gradient Variances for Out-of-distribution Generalization

### PR DESCRIPTION
**Pull request by alexrame**
We propose a new regularization - named Fishr - that enforces domain invariance in the space of the gradients of the loss function: specifically, it tackles the domain-level variances of gradients across the training domains.
Compared to existing methods on DomainBed, Fishr performs best in the 'Oracle' model selection and third in the 'Training' model selection. Moreover, Fishr performs consistently better than ERM.
The paper is on [arXiv](https://arxiv.org/abs/2109.02934).

### CODE REVIEW

**Comment by lopezpaz** [comment]
Beautiful, thank you!
